docs/src/main/asciidoc/maven-tooling.adoc

```diff
@@ -640,3 +631,73 @@ your default testing database.
+
+Here is a list of system properties the Quarkus bootstrap Maven resolver checks during its initialization.
+
+[cols=3*,options="header"]
+|===
+| Property name
+| Default Value
+| Description
+
+| `maven.home`
+| `MAVEN_HOME` envvar
+| The Maven home dir is used to resolve the global settings file unless it was explicitly provided on the command line with the `-gs` argument
+
+| `maven.settings`
+| `~/m2/settings.xml`
+| Unless the custom settings file has been provided with the `-s` argument, this property can be used to point the resolver to a custom Maven settings file
+
+| `maven.repo.local`
+| `~/m2/repository`
+| This property could be used to configure a custom local Maven repository directory, if it is different from the default one and the one specified in the `settings.xml`
+
+| `maven.top-level-basedir`
+| none
+| This property may be useful to help the Maven resolver identify the top-level Maven project in the workspace. By default, the Maven resolver will be discovering a project's workspace by navigating the parent-module POM relationship. However there could be project layouts that are using an aggregator module which isn't appearing as the parent for its modules. In this case, this property will help the Quarkus Maven resolver to properly discover the workspace.
+
+| `quarkus.bootstrap.effective-model-builder`
+| `false`
+| By default, the Quarkus Maven resolver is reading project's POMs directly when discovering the project's layout. While in most cases it works well enough and relatively fast, reading raw POMs has its limitation. E.g. if a POM includes modules in a profile, these modules will not be discovered. This system property enables project's layout discovery based on the effective POM models, that are properly interpolated, instead of the raw ones. The reason this option is not enabled by default is it may appear to be significantly more time consuming that could increase, e.g. CI testing times. Until there is a better approach found that could be used by default, projects that require it should enable this option.
+
+|===
+
+These system properties above could be added to, e.g., a `surefire` and/or `failsafe` plugin configuration as
+[source,xml,subs=attributes+]
+----
+<project>
+  [...]
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <version>${surefire-plugin.version}</version>
+        <configuration>
+          <systemPropertyVariables>
+            <maven.home>${maven.hom}</maven.home> <1>
+            <maven.repo.local>${settings.localRepository}</maven.repo.local> <2>
+            <maven.settings>${session.request.userSettingsFile.path}</maven.settings> <3>
+            <maven.top-level-basedir>${session.topLevelProject.basedir.absolutePath}</maven.top-level-basedir> <4>
```

Comment:
Doesn't seem to provide the correct dir when used with `-f`:
```
$ mvn help:evaluate -Dexpression=session.topLevelProject.basedir.absolutePath -q -DforceStdout -f integration-tests/flyway/
/home/famod/proj/quarkus/integration-tests/flyway
```

```
$ mvn help:evaluate -Dexpression=maven.multiModuleProjectDirectory -q -DforceStdout -f integration-tests/flyway/
/home/famod/proj/quarkus
```

Reply:
Well, that's what Maven reports as the top-level project. Right?

Reply:
When you use `-f` that's what you are asking for, right?

Reply:
But I also see your point.

Reply:
If I use the reproducer from the original report, either property works. So, I'm wondering whether `maven.multiModuleProjectDirectory` is a better choice.

Reply:
I could mention both properties in the doc too.

Reply:
> If I use the reproducer from the original report, either property works.

But not with `-f`, right? Even `maven.multiModuleProjectDirectory` doesn't work there, which is expected since there is no `.mvn` directory.

> So, I'm wondering whether `maven.multiModuleProjectDirectory` is a better choice.

Yeah, seems better, but not perfect. I think if you go for that you should also add a note to the docs w.r.t. `.mvn`.

Reply:
> > If I use the reproducer from the original report, either property works.
> 
> But not with `-f`, right? Even `maven.multiModuleProjectDirectory` doesn't work there, which is expected since there is no `.mvn` directory.

I tried both with `-f`

> > So, I'm wondering whether `maven.multiModuleProjectDirectory` is a better choice.
> 
> Yeah, seems better, but not perfect. I think if you go for that you should also add a note to the docs w.r.t. `.mvn`.

I added https://github.com/quarkusio/quarkus/pull/17351/files#diff-22914b1e1fa444916b50c249fd5e9ac895c5953f40602207dd83cb0f59c10e5aR705

Reply:
I would rather reference some Maven docs about how the multi-module project dir is resolved instead of writing it myself. Do you by chance know where it's documented?

Reply:
AFAIK, it's not really documented. And it seems some Maven maintainers don't like users taking advantage of it:
https://lists.apache.org/list.html?users@maven.apache.org:2021-2

Reply:
PS: Link fixed ⬆️.
PPS: I use that property in my projects regardless of that controversy. 🤷

Reply:
Added a note that it's not an officially supported property.

Reply:
Yeah ok. I mean it's out there, people are using it, for good reasons!
But to be on the safe side for a big public framework like Quarkus, this defensive approach makes sense.

Reply:
Thanks @famod